### PR TITLE
Support incremental updates for UPDATE queries

### PR DESCRIPTION
**Pull request by tomekl007**
Adds a new `@cql_increment` directive that allows customers to use CQL increment operations on Counter, List, and Set via Grapqhl-first API, for example:
```
type Mutation {
  updateCounter(
    k: Int
    cInc: Int @cql_increment(field: "c")
  ): Boolean
  
  appendList(
    k: Int
    l: [Int] @cql_increment
  ): Boolean

  prependList(
    k: Int
    l: [Int] @cql_increment(prepend: true)
  ): Boolean
}
```
todo: 
- [x] support for counter 
- [x] support for append list
- [x] support for prepend list
- [x] validation

### CODE REVIEW

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/DirectiveModelsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+          "Operation %s: could not find field %s in type %s",
+          operationName, fieldName, entity.getGraphqlName());
+    }
+    return result;
+  }
+
+  static class BuildState {
+
+    private boolean foundErrors = false;
+    private boolean isWhereCondition = false;
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/DirectiveModelsBuilder.java:270]: Hrmm, not a big fan of this, although I understand that you did it to allow helper methods to modify the booleans.

The whole class is kind of hard to follow now. Maybe breaking down the three cases (select/delete/update) into submethods would help...

**Review by olim7t — CHANGES_REQUESTED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/UpdateModelBuilder.java [hunk 2]
```diff
@@ -117,7 +120,26 @@ MutationModel build() throws SkipException {
...
+  }
+
+  private Optional<IncrementModel> getAndValidate(List<IncrementModel> incrementModel)
+      throws SkipException {
+
+    if (incrementModel.size() > 1) {
+      invalidMapping(
+          "The %s directive can be set only on one field, but it was set on %s fields.",
+          CqlDirectives.INCREMENT, incrementModel.size());
+      throw SkipException.INSTANCE;
```
[the remark this answers is not in the document]
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/UpdateModelBuilder.java]: done

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    }
+  }
+
+  private void validateCounter(TypeName fieldInputType) throws SkipException {
+    String typeName = fieldInputType.getName();
+    if (typeName.equalsIgnoreCase(COUNTER_TYPE_NAME)) {
+      // counter graph-ql field can be BIGINT or INT
+      checkArgumentIsAnyOfTypes(
+          Arrays.asList(CqlScalar.BIGINT.getGraphqlType(), Scalars.GraphQLInt));
+    }
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:103]: This should fail if the type is not `Counter`.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/UpdateModel.java [hunk 4]
```diff
@@ -82,6 +85,10 @@ public ReturnType getReturnType() {
     return responsePayload;
   }
 
+  public List<IncrementModel> getIncrementModel() {
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/UpdateModel.java:88]:
```suggestion
  public List<IncrementModel> getIncrementModels() {
```

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/UpdateModelBuilder.java [hunk 5]
```diff
@@ -94,14 +95,16 @@ MutationModel build() throws SkipException {
       entity = entityFromFirstArgument.get();
       whereConditions = entity.getPrimaryKeyWhereConditions();
       ifConditions = Collections.emptyList();
+      incrementModels = Collections.emptyList();
     } else {
       entity = entityFromDirective(cqlUpdateDirective, "update", CqlDirectives.UPDATE);
-      ConditionModels conditions =
-          new ConditionModelsBuilder(operation, OperationType.UPDATE, entity, entities, context)
+      ArgumentDirectiveModels directives =
+          new DirectiveModelsBuilder(operation, OperationType.UPDATE, entity, entities, context)
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/UpdateModelBuilder.java:102]: Since we renamed the class, the builder should be renamed as well.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    Type<?> argumentType = TypeHelper.unwrapNonNull(argument.getType());
+
+    boolean hasProperType =
+        scalarTypes.stream()
+            .anyMatch(
+                type -> {
+                  if (argumentType instanceof TypeName) {
+                    return type.getName().equals(((TypeName) argumentType).getName());
+                  }
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:131]: Nit: you can extract `argumentType.getName()` outside of the loop.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+    Type<?> fieldInputType =
+        toInput(field.getGraphqlType(), argument, entity, field, entities, operationName);
+    // if it is non-list type
+    if (fieldInputType instanceof TypeName) {
+      validateCounter((TypeName) fieldInputType);
+      validatePrepend(field, prepend);
+    } else {
+      // otherwise it must be a list
+      checkArgumentIsListOf(argument, entity, entities, operationName, field);
+    }
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:94]: We should also fail if `prepend` is set and the column is a a CQL set.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+    Type<?> fieldInputType =
+        toInput(field.getGraphqlType(), argument, entity, field, entities, operationName);
+    // if it is non-list type
+    if (fieldInputType instanceof TypeName) {
+      validateCounter((TypeName) fieldInputType);
+      validatePrepend(field, prepend);
+    } else {
+      // otherwise it must be a list
+      checkArgumentIsListOf(argument, entity, entities, operationName, field);
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:93]: This not the right check.

What that method does is check that the argument _is a list of_ the field type (e.g. field = `Int` => argument = `[Int]`).

What we want is to check that both the argument and the field are lists of the same element type. For example:
```graphql
    type List {
      id: ID
      l: [Int]
    }
    type Mutation {
      updateList(
        id: ID
        lSuffix: [Int] @cql_increment(field: "l")
      ): Boolean @cql_update(targetEntity: "List")
    }
```

PS: I tried to deploy the above and to my surprise it worked. Turns out that there is a bug in `checkArgumentIsListOf`. I'll open a separate ticket to fix it.

**Review by olim7t — CHANGES_REQUESTED** [review]

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java (hunk 6 above)
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:131]: To use it in lambda we would still need to copy it:
```
 String argumentTypeName = null;
    if (argumentType instanceof TypeName) {
      argumentTypeName = ((TypeName) argumentType).getName();
    }
    String finalArgumentTypeName = argumentTypeName;
```
so I am not sure if extracting it is a good choice.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java (hunk 7 above)
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:94]: Is it even possible to create a set using the GrapqhQL API?
I don't see any test covering that case.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java (hunk 8 above)
Comment (tomekl007) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:93]: fixed in:
https://github.com/stargate/stargate/pull/1015

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java (hunk 6 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java]: Addressed in #e9b4f103

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java (hunk 7 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java]: The CQL type is known at this stage, via `field.getCqlType()`. Fixed in 3e2b33f0.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java (hunk 8 above)
Comment (olim7t) [reply graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java]: Even after fixing the method, we still have to change the call here, as I said it's not the right check. Also addressed in 3e2b33f0.

File: graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.*;
+
+public class IncrementModelBuilder extends ArgumentDirectiveModelBuilderBase<IncrementModel> {
+
+  // The types that we allow for an argument that represents an increment to a counter:
+  private static final List<String> COUNTER_INCREMENT_TYPES =
+      ImmutableList.of(
+          Scalars.GraphQLInt.getName(),
+          CqlScalar.BIGINT.getGraphqlType().getName(),
+          CqlScalar.COUNTER.getGraphqlType().getName());
```
Comment (olim7t) [inline graphqlapi/src/main/java/io/stargate/graphql/schema/graphqlfirst/processor/IncrementModelBuilder.java:37]: I've added `Counter`, it feels natural to use the same type as the field, and this didn't require any additional code.

**Review by olim7t — APPROVED** [review]
I've addressed a couple of remaining items, and attempted a refactoring of the builder in 43e418e9.
